lib/engine/game/g_18_co.rb

```diff
@@ -266,9 +276,44 @@ def action_processed(action)
         case action
         when Action::BuyCompany
           mine_update_text(action.entity) if action.company == imc && action.entity.corporation?
+        when Action::PlaceToken
+          remove_corporations_if_no_home(action.city) if @phase.status.include?('closable_corporations')
+        when Action::Par
+          rereserve_home_station(action.corporation) if @phase.status.include?('closable_corporations')
         end
       end
 
+      def remove_corporations_if_no_home(city)
+        tile = city.tile
+
+        return unless tile_has_max_cities(tile)
+
+        @corporations.dup.reject(&:ipoed).each do |corp|
```

Comment:
there's no reason to call dup, reject already does that

Reply:
or if you're already using the each next pattern, then do

dup.each do 

next if ipoed